Bluetooth: Host: Various fixes for L2CAP collision mitigation

hermabe opened the pull request:
This fixes various issues with the L2CAP collision mitigation as found by @alwa-nordic in https://github.com/zephyrproject-rtos/zephyr/pull/43037#discussion_r813707314.

- The number of channels requested in the retry was wrong in some cases, using the number of channels from the incoming request instead of outgoing request.
- If the peer disconnected between the time the retry was scheduled and when it was executed, a null pointer would be dereferenced.
- Collisions were detected for other PSMs than the EATT PSM.
- Two outgoing connection requests to the same peer being rejected with "No resources" would trigger the retry. The retry shall only be triggered if both an incoming and an outgoing request are rejected with "No resources".

hermabe committed:
Bluetooth: Host: Fix L2CAP collision mitigation

There were some errors in the detection of ECRED connection collisions,
so the retry was triggered incorrectly.

The number of channels requested in the retry was wrong in some cases,
using the number of channels from the received request instead of the
sent request.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1652,7 +1656,7 @@ static void le_ecred_conn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
 
 	if (ecred_cb && ecred_cb->ecred_conn_rsp) {
-		ecred_cb->ecred_conn_rsp(conn, result, attempted, succeeded);
+		ecred_cb->ecred_conn_rsp(conn, result, attempted, succeeded, psm);
```
Since the `ecred_conn_rsp` is an internal callback we haven't really documented it which is fine I suppose, but the it may not be clear that the `psm` is the `psm` of the last channel handled. 

Since we are working on multiple channels, perhaps the callback should contain an array of { psm, success}, e.g.

```c
struct ecred_conn_rsp_res {
    bool success;
    uint16_t psm;
}

void (*ecred_conn_rsp)(struct bt_conn *conn, uint16_t result, struct ecred_conn_rsp_res rsp_res[], size_t rsp_cnt);
```

What do you think? Or is the PSM of the last channel all we need?

hermabe replied:
All the channels have the same `ident`, so they come from the same request and thus have the same `psm`.

Though I don't really like these callbacks so I am open to changing them.

Thalley replied:
Ah, in that case the `psm` assignment should probably be moved outside of the `while` loops (no reason to assign it multiple times if it will be the same). Perhaps we should add an `__ASSERT` to verify that the channel do indeed have the same PSM then though?

hermabe replied:
Maybe better to add the assert in `l2cap_ecred_conn_req()` instead/as well? There the PSM of the first channel is used for the request.

hermabe committed:
Bluetooth: Host: Assert that all channels in request have the same PSM

The PSM of the first channel in the connection request is used, but the
PSM value is used for all of them on the receiving side.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -105,9 +105,14 @@ struct bt_att {
+		uint16_t prev_eatt_conn_rsp_result;
+		uint16_t prev_eatt_conn_req_result;
+		uint8_t prev_eatt_conn_req_missing_chans;
+	} conn_context;
```
Why not just call this `eatt` instead of `conn_context`? You could then also remove any mentions of "eatt" in the struct member names, since it's already implied.
